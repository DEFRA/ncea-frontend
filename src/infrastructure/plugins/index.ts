export { customHapiPino } from './logger';
export { customHapiRoutes } from './router';
export { customHapiViews } from './views';

