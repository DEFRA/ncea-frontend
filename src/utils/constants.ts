import { IFilterOptions } from '../interfaces/searchPayload.interface';
import { TabOptions } from '../interfaces/detailsTab.interface';

export const webRoutePaths = {
  home: '/',
  results: '/search',
  guidedDateSearch: '/date-search',
  geographySearch: '/coordinate-search',
  intermediate: '/intermediate',
  getMapResults: '/map-results',
  getMapFilters: '/map-filters',
  filterResourceType: '/resource-type-filter',
  filterStudyPeriod: '/study-period-filter',
  sortResults: '/sort-results',
  help: '/help',
  accessibilityStatement: '/accessibility-statement',
  termsAndConditions: '/terms-conditions',
  privacyPolicy: '/privacy-policy',
  cookiePolicy: '/cookie-policy',
};

export const elasticSearchAPIPaths = {
  searchPath: '_search',
  countPath: '_count',
};

export const formKeys = {
  dateQuestionnaire: 'date-questionnaire',
};

export const formIds = {
  quickSearchFID: 'keyword',
  dataQuestionnaireFID: 'date',
  geographyQuestionnaireFID: 'extent',
};

export const showMoreText: string = 'Show more';
export const showLessText: string = 'Show less';
export const maxWords: number = 100;
export const geoNetworkIndex: string = 'gn-records';
export const mapResultMaxCount: number = 10000;
export const requiredFieldsForMap: string[] = [
  'resourceTitleObject',
  'contactForResource',
  'resourceAbstractObject',
  'geom',
  'resourceIdentifier',
  'resourceType',
  'resourceTemporalExtentDetails',
  'ownerOrgForResourceObject',
  'originatorOrgForResourceObject',
  'custodianOrgForResourceObject',
  'distributorOrgForResourceObject',
  'pointOfContactOrgObject',
  'linkUrl',
  'OrgNceaClassifiers'
];

export const quickSearchTargetFields: string[] = [
  'metadataIdentifier',
  'resourceTitleObject.default',
  'resourceAltTitleObject.default',
  'resourceAbstractObject.default',
  'lineageObject.default',
  'tag.default',
  'cl_topic.default',
  'resourceType',
  'supplementalInformationObject.default',
];

export const guidedSearchSteps = {
  date: 'date',
  coordinate: 'extent',
};

export const queryParamKeys = {
  quickSearch: 'q',
  fromDateDay: 'fdd',
  fromDateMonth: 'fdm',
  fromDateYear: 'fdy',
  toDateDay: 'tdd',
  toDateMonth: 'tdm',
  toDateYear: 'tdy',
  coordinates: 'ext',
  north: 'nth',
  south: 'sth',
  east: 'est',
  west: 'wst',
  count: 'cnt',
  page: 'pg',
  rowsPerPage: 'rpp',
  sort: 'srt',
  journey: 'jry',
  resourceType: 'rty',
  startYear: 'sy',
  toYear: 'ty',
};

export const uniqueResourceTypesKey: string = 'unique_resource_types';
export const startYearRangeKey: string = 'start_year_range';
export const toYearRangeKey: string = 'to_year_range';
export const yearRange: string = 'year';

export const resourceTypeFilterField = 'resourceTypeFilter';
export const studyPeriodFilterField = 'studyPeriodFilter';

export const defaultFilterOptions: IFilterOptions = [
  {
    key: uniqueResourceTypesKey,
    needCount: true,
    propertyToRead: 'key',
    hasBucket: true,
    isTerm: true,
  },
  {
    key: yearRange,
    needCount: false,
    propertyToRead: 'value_as_string',
    hasBucket: false,
    isDate: true,
  },
];

export const detailsTabOptions: TabOptions = {
  general: {
    Abstract: 'content',
    'Study periods': 'studyPeriod',
    'Topic categories': 'topicCategories',
    Keywords: 'keywords',
    'Resource languages': 'language',
  },
  access: {
    'File Identifier': 'ncea_catalogue_number',
    'Resource Identifier': 'host_catalogue_number',
    'Coupled resources': 'host_catalogue_entry',
    'Resource type': 'resource_type_and_hierarchy',
    'Hierarchy level name': 'hierarchy_level',
    'Resource locators': 'resource_locators',
    'Catalouge entry': 'ncea_catalogue_number',
    'NCEA search service catalogue number': 'host_service_catalogue_number',
    'Parent ID': 'ncea_group_reference',
    'Metadata standard': 'metadata_standard',
    'Project number': 'project_number',
    'Metadata language': 'metadata_language',
  },
  quality: {
    'Date of publication': 'publicationInformation',
    'Date of creation': 'creationInformation',
    'Date of last revision': 'revisionInformation',
    'Metadata date': 'metadataDate',
    Lineage: 'lineage',
    Conformity: 'conformity',
    'Additional information': 'additionalInformation',
  },
  geography: {
    'Spatial data service': 'spatialDataService',
    'Spatial representation service': 'spatialRepresentationService',
    'Spatial referencing system': 'spatialReferencingSystem',
    'Geographic locations': 'geographicLocations',
    'Geographic boundary': 'geographicBoundaryHtml',
    map: 'geographicBoundary',
    'Vertical extent<br /><span>(Meters above sea level)</span>': 'verticalExtent',
    'Spatial resolution': 'samplingResolution',
  },
  license: {
    'Limitations on public access': 'limitation_on_public_access',
    'License constraints': 'license_constraints',
    'Distribution formats': 'available_formats',
    'Frequency of update': 'frequency_of_update',
    'Character encoding': 'character_encoding',
  },
};

export const pageTitles = {
  home: 'NCEA Search Service Home',
  date: 'NCEA Guided Search - Study Period',
  geography: 'NCEA Guided Search - Geographic Data',
  results: 'NCEA Search Results Summary',
  generalTab: 'NCEA Catalogue Detail - General',
  accessTab: 'NCEA Catalogue Detail - Access',
  naturalCapitalTab: 'NCEA Catalogue Detail - Natural Capital',
  qualityTab: 'NCEA Catalogue Detail - Quality',
  geographyTab: 'NCEA Catalogue Detail - Geography',
  governanceTab: 'NCEA Catalogue Detail - Governance',
  licenseTab: 'NCEA Catalogue Detail - License',
  help: 'Help',
  accessibility: 'Accessibility statement',
  termsAndConditions: 'Terms and conditions',
  privacyPolicy: 'Privacy policy',
  cookiePolicy: 'Cookies',
};
