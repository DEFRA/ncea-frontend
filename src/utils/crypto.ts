import * as crypto from 'crypto';

const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
  modulusLength: 2048,
  publicKeyEncoding: {
    type: 'spki',
    format: 'pem',
  },
  privateKeyEncoding: {
    type: 'pkcs8',
    format: 'pem',
  },
});

const encryptData = (data: string): string => {
  const encryptedBuffer = crypto.publicEncrypt(
    {
      key: publicKey,
      padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
      oaepHash: 'sha256',
    },
    Buffer.from(data),
  );

  return encryptedBuffer.toString('base64');
};

const decryptData = (encryptedData: string): string => {
  const decryptedBuffer = crypto.privateDecrypt(
    {
      key: privateKey,
      padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
      oaepHash: 'sha256',
    },
    Buffer.from(encryptedData, 'base64'),
  );

  return decryptedBuffer.toString();
};

export { encryptData, decryptData };
