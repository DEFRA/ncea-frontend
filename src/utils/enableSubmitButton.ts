import { IFormValidatorOptions } from '../interfaces/guidedSearch.interface';

const enableSubmitButton = async (formValidatorOptions: IFormValidatorOptions) => {
  const form = document.getElementById(formValidatorOptions.formId) as HTMLFormElement;
  const submitButton = document.getElementById(formValidatorOptions.submitButtonId) as HTMLButtonElement;
  const fields = Array.from(form.querySelectorAll('input[type="text"]')) as HTMLInputElement[];

  const checkFields = (): void => {
    const hasValue = fields.some((field) => field.value.trim() !== '');
    submitButton.disabled = !hasValue;
  };

  fields.forEach((field) => {
    field.addEventListener('input', checkFields);
  });

  checkFields();
};

const injectDynamicEnablingScript = (formValidatorOptions: IFormValidatorOptions) => {
  const script = `<script>
      (${enableSubmitButton.toString()})(${JSON.stringify(formValidatorOptions)});
    </script>`;
  return script;
};

export { enableSubmitButton, injectDynamicEnablingScript };
