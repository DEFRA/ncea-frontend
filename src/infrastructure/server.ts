'use strict';

import { environmentConfig } from '../config/environmentConfig';
import inert from '@hapi/inert';
import vision from '@hapi/vision';
import Hapi, { Server } from '@hapi/hapi';
import logger from '../infrastructure/plugins/appinsights-logger';

import { customHapiPino, customHapiRoutes, customHapiViews } from './plugins/index';

// Create the hapi server
const server: Server = Hapi.server({
  host: process.env.HOST ?? '0.0.0.0',
  port: environmentConfig.env !== 'test' ? environmentConfig.port : 4000,
  routes: {
    validate: {
      options: {
        abortEarly: false,
      },
    },
  },
});

const initializeServer = async (): Promise<Server> => {
  // Register vendors plugins
  await server.register([inert, vision]);

  // Register the custom plugins
  await server.register([customHapiViews, customHapiRoutes, customHapiPino]);

  await server.initialize();
  return server;
};

const startServer = async (): Promise<Server> => {
  logger.info('Initializing server');
  await server.start();  
  console.log(`Server running at: ${server.info.uri}`);
  return server;
};

export { initializeServer, startServer };
