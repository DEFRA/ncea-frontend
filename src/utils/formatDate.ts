import { months } from './constants';

const formatDate = (dateString: string, isFormatting: boolean = false): string => {
  const isValidDate = (dateString: string): boolean => {
    const date = new Date(dateString);
    return !isNaN(date.getTime());
  };

  if (!isValidDate(dateString)) {
    return '';
  }

  const date = new Date(dateString);
  let day: number | string = date.getDate();
  const month: string = months[date.getMonth()]!;
  const year: number = date.getFullYear();
  let hours: number | string = date.getHours();
  let minutes: number | string = date.getMinutes();
  day = day < 10 ? '0' + day : day;

  let formattedDate = isFormatting ? `${day}-${month}-${year}` : `${day} ${month} ${year}`;

  if ((hours !== 0 || minutes !== 0) && !isFormatting) {
    hours = hours < 10 ? '0' + hours : hours;
    minutes = minutes < 10 ? '0' + minutes : minutes;
    formattedDate += ` @${hours}${minutes}`;
  }

  return formattedDate;
};

export { formatDate };
