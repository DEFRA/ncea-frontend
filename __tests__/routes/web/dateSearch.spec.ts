/**
 * @jest-environment jsdom
 */

'use strict';

import { Server, ServerInjectResponse } from '@hapi/hapi';

import { initializeServer } from '../../../src/infrastructure/server';
import { webRoutePaths } from '../../../src/utils/constants';

jest.mock('../../../src/infrastructure/plugins/appinsights-logger', () => ({
  logger: jest.fn(),
}));

jest.mock('../../../src/utils/keyvault', () => ({
  getSecret: jest.fn(),
}));

describe('Guided Search - Date Questionnaire Screen', () => {
  let server: Server;
  let response: ServerInjectResponse<object>;
  let document: Document;

  beforeAll((done) => {
    initializeServer().then(async (s: Server) => {
      server = s;

      const options = {
        method: 'GET',
        url: webRoutePaths.guidedDateSearch,
      };

      response = await server.inject(options);
      const rawHTML = response.payload;
      const parser = new DOMParser();
      document = parser.parseFromString(rawHTML, 'text/html');
      done();
    });
  });

  afterAll((done) => {
    server.stop().then(() => done());
  });

  describe('Guided Search > Date Questionnaire > Snapshot verification', () => {
    it('should match the date questionnaire screen snapshot', async () => {
      expect(response.payload).toMatchSnapshot();
    });
  });

  describe('Guided Search > Date Questionnaire > Check /results route status code', () => {
    it('should /guided-search route works with status code 200', async () => {
      expect(response.statusCode).toEqual(200);
    });
  });

  describe('Guided Search > Date Questionnaire > Back link block', () => {
    describe('Back link classes', () => {
      it('renders the back-link class', async () => {
        expect(document.querySelector('.govuk-back-link')).toBeTruthy();
      });
    });

    describe('Back link items', () => {
      it('should render 1 item', async () => {
        expect(document.querySelectorAll('.govuk-back-link')?.length).toEqual(
          1,
        );
      });
    });

    describe('Back link  item options', () => {
      it('should render back link to date search page', async () => {
        const anchor = document.querySelector('.govuk-back-link');
        expect(anchor?.tagName.toLowerCase()).toBe('a');
        expect(anchor?.getAttribute('href')).toEqual(webRoutePaths.home);
        expect(anchor?.textContent?.trim()).toEqual('Back');
      });
    });
  });

  describe('Guided Search > Date Questionnaire block', () => {
    describe('Date Questionnaire block classes', () => {
      it('renders govuk-grid-column-full class', async () => {
        expect(document.querySelector('.govuk-grid-column-full')).toBeTruthy();
      });
    });

    describe('Date Questionnaire block heading', () => {
      it('should render the search container heading', async () => {
        expect(
          document
            ?.querySelectorAll('.govuk-heading-m')?.[0]
            ?.textContent?.trim(),
        ).toBe('When was the data or information collected?');
        expect(
          document
            ?.querySelectorAll('.govuk-heading-m')?.[1]
            ?.textContent?.trim(),
        ).toBe('From');
        expect(
          document
            ?.querySelectorAll('.govuk-heading-m')?.[2]
            ?.textContent?.trim(),
        ).toBe('To');
      });
    });

    describe('Date questionnaire form', () => {
      it('should render the form', async () => {
        const rowElements = document?.querySelectorAll('.govuk-grid-row');
        if (rowElements.length > 1) {
          const formElement = rowElements[1]?.firstElementChild;
          expect(formElement?.tagName.toLowerCase()).toBe('form');
        }
      });
    });

    describe('date questionnaire input field', () => {
      it('should renders with from date day id', () => {
        const inputElement = document?.querySelector('#day');
        expect(inputElement).toBeTruthy();
      });

      it('should renders with from date month id', () => {
        const inputElement = document?.querySelector('#month');
        expect(inputElement).toBeTruthy();
      });

      it('should renders with from date year id', () => {
        const inputElement = document?.querySelector('#year');
        expect(inputElement).toBeTruthy();
      });

      it('should renders with to date day id', () => {
        const inputElement = document?.querySelector('#day');
        expect(inputElement).toBeTruthy();
      });

      it('should renders with to date month id', () => {
        const inputElement = document?.querySelector('#month');
        expect(inputElement).toBeTruthy();
      });

      it('should renders with to date year id', () => {
        const inputElement = document?.querySelector('#year');
        expect(inputElement).toBeTruthy();
      });

      it('should renders a button', () => {
        const buttonElement = document?.querySelector(
          'button[data-module="govuk-button"]',
        );
        expect(buttonElement).toBeTruthy();
      });
    });
  });
});
