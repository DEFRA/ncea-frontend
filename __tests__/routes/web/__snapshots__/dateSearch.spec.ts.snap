// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Guided Search - Date Questionnaire Screen Guided Search > Date Questionnaire > Snapshot verification should match the date questionnaire screen snapshot 1`] = `
"<!DOCTYPE html>
<html lang="en" class="govuk-template app-html-class">
  <head>
    <meta charset="utf-8">
    <title>
  Natural Capital Search Service - GOV.UK
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover">
    <meta name="theme-color" content="#0b0c0c">

    
  
      <link rel="icon" sizes="48x48" href="/assets/images/favicon.ico">
      <link rel="icon" sizes="any" href="/assets/images/favicon.svg" type="image/svg+xml">
      <link rel="mask-icon" href="/assets/images/govuk-icon-mask.svg" color="#0b0c0c">
      <link rel="apple-touch-icon" href="/assets/images/govuk-icon-180.png">
      <link rel="manifest" href="/assets/manifest.json">
    


    
  <link rel="stylesheet" href="/assets/application.css">

    
  </head>
  <body class="govuk-template__body app-body-class">
    <script>document.body.className += ' js-enabled' + ('noModule' in HTMLScriptElement.prototype ? ' govuk-frontend-supported' : '');</script>
    

    
      <a href="#main-content" class="govuk-skip-link" data-module="govuk-skip-link">Skip to main content</a>

    

    
  <header class="header-container">
    <div class="header-div-container header-main">
      <div class="site-width-container">
        <div class="header-logo-container"><img alt="Department for Environment Food &amp; Rural Affairs Logo" src="/assets/images/defra_logo.svg" width="52" height="42" decoding="async" data-nimg="1" class="defra-logo" loading="lazy" style="color:transparent">
          <a target="_blank" class="header-logo-link logo-link" href="https://www.gov.uk/government/organisations/department-for-environment-food-rural-affairs">
            <span class="defra-logo-text">Department for Environment <br> Food &amp; Rural Affairs</span></a>
          <span class="govuk-header__service-name">Natural Capital Search Service</span></div>
      </div>
    </div>
  </header>


    
  
      <div class="govuk-width-container">
        
  
  <div class="govuk-phase-banner">
  <p class="govuk-phase-banner__content">
    <strong class="govuk-tag govuk-phase-banner__content__tag">
      Beta
    </strong>
    <span class="govuk-phase-banner__text">
      This is a new service – your <a class="govuk-link" href="#">feedback</a> will help us to improve it.
    </span>
  </p>
</div>


  <div class="govuk-breadcrumbs">
  <ol class="govuk-breadcrumbs__list">

  
    <li class="govuk-breadcrumbs__list-item">
      <a class="govuk-breadcrumbs__link" href="/" data-do-storage-reset="">Home</a>
    </li>
  

  
    <li class="govuk-breadcrumbs__list-item">
      <a class="govuk-breadcrumbs__link" href="#">Questionnaire search</a>
    </li>
  

  </ol>
</div>


        <main class="govuk-main-wrapper app-main-class" id="main-content" role="main">
          
  <div class="govuk-grid-row">
    <form method="POST" id="date-search" action="/date-search" data-do-browser-storage>
      <div class="govuk-grid-column-two-thirds">
        <h1 class="govuk-heading-m">
          When was the data or information collected?
        </h1>
        

<h2 class="govuk-heading-m">
  From
</h2>




  
<div class="govuk-form-group">

  <fieldset class="govuk-fieldset" role="group" aria-describedby="from-date-hint">
    
    <legend class="govuk-fieldset__legend">
    
      You must enter a year, for example, 2007
    
    </legend>
    
  
    <div id="from-date-hint" class="govuk-hint">
      You can choose to enter a day or month, for example, 27 3 2007
    </div>
  
  
    <div class="govuk-date-input" id="from-date">
      
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="from-date-day">
            Day
          </label>
        
        
          <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="from-date-day" name="from-date-day" type="text" inputmode="numeric">
        
        </div>
      </div>
    
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="from-date-month">
            Month
          </label>
        
        
          <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="from-date-month" name="from-date-month" type="text" inputmode="numeric">
        
        </div>
      </div>
    
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="from-date-year">
            Year
          </label>
        
        
          <input class="govuk-input govuk-date-input__input govuk-input--width-4" id="from-date-year" name="from-date-year" type="text" inputmode="numeric">
        
        </div>
      </div>
    
    </div>
  
  </fieldset>

</div>

        

<h2 class="govuk-heading-m">
  To
</h2>




  
<div class="govuk-form-group">

  <fieldset class="govuk-fieldset" role="group" aria-describedby="to-date-hint">
    
    <legend class="govuk-fieldset__legend">
    
      You must enter a year, for example, 2007
    
    </legend>
    
  
    <div id="to-date-hint" class="govuk-hint">
      You can choose to enter a day or month, for example, 27 3 2007
    </div>
  
  
    <div class="govuk-date-input" id="to-date">
      
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="to-date-day">
            Day
          </label>
        
        
          <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="to-date-day" name="to-date-day" type="text" inputmode="numeric">
        
        </div>
      </div>
    
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="to-date-month">
            Month
          </label>
        
        
          <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="to-date-month" name="to-date-month" type="text" inputmode="numeric">
        
        </div>
      </div>
    
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="to-date-year">
            Year
          </label>
        
        
          <input class="govuk-input govuk-date-input__input govuk-input--width-4" id="to-date-year" name="to-date-year" type="text" inputmode="numeric">
        
        </div>
      </div>
    
    </div>
  
  </fieldset>

</div>

        
  
    
  

<button type="submit" class="govuk-button" data-module="govuk-button">
  Show results
</button>

      </div>
    </form>
  </div>

        </main>
      </div>
    


    
  <footer class="govuk-footer" role="contentinfo">
  <div class="govuk-width-container">
    
    <div class="govuk-footer__meta">
      <div class="govuk-footer__meta-item govuk-footer__meta-item--grow">
        
          <h2 class="govuk-visually-hidden">Support links</h2>
          
            <ul class="govuk-footer__inline-list">
              
                <li class="govuk-footer__inline-list-item">
                  <a class="govuk-footer__link" href="#">
                    Help
                  </a>
                </li>
              
                <li class="govuk-footer__inline-list-item">
                  <a class="govuk-footer__link" href="#">
                    Privacy
                  </a>
                </li>
              
                <li class="govuk-footer__inline-list-item">
                  <a class="govuk-footer__link" href="#">
                    Cookies
                  </a>
                </li>
              
                <li class="govuk-footer__inline-list-item">
                  <a class="govuk-footer__link" href="#">
                    Accessibility statement
                  </a>
                </li>
              
            </ul>
          
          
        
        <svg
          aria-hidden="true"
          focusable="false"
          class="govuk-footer__licence-logo"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 483.2 195.7"
          height="17"
          width="41"
        >
          <path
            fill="currentColor"
            d="M421.5 142.8V.1l-50.7 32.3v161.1h112.4v-50.7zm-122.3-9.6A47.12 47.12 0 0 1 221 97.8c0-26 21.1-47.1 47.1-47.1 16.7 0 31.4 8.7 39.7 21.8l42.7-27.2A97.63 97.63 0 0 0 268.1 0c-36.5 0-68.3 20.1-85.1 49.7A98 98 0 0 0 97.8 0C43.9 0 0 43.9 0 97.8s43.9 97.8 97.8 97.8c36.5 0 68.3-20.1 85.1-49.7a97.76 97.76 0 0 0 149.6 25.4l19.4 22.2h3v-87.8h-80l24.3 27.5zM97.8 145c-26 0-47.1-21.1-47.1-47.1s21.1-47.1 47.1-47.1 47.2 21 47.2 47S123.8 145 97.8 145"
          />
        </svg>
        <span class="govuk-footer__licence-description">
          
            All content is available under the
            <a
              class="govuk-footer__link"
              href="https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/"
              rel="license"
            >Open Government Licence v3.0</a>, except where otherwise stated
          
        </span>
      </div>
      <div class="govuk-footer__meta-item">
        <a
          class="govuk-footer__link govuk-footer__copyright-logo"
          href="https://www.nationalarchives.gov.uk/information-management/re-using-public-sector-information/uk-government-licensing-framework/crown-copyright/"
        >© Crown copyright</a>
      </div>
    </div>
  </div>
</footer>



    
  
  <script type="module" src="/assets/all.js"></script>
  <script type="module">
    import {initAll} from '/assets/all.js'
    initAll()
  </script>
  <script type="text/javascript">
    !(function (cfg) {
      function e() {
        cfg.onInit && cfg.onInit(i)
      }
      var S,
        u,
        D,
        t,
        n,
        i,
        C = window,
        x = document,
        w = C.location,
        I = "script",
        b = "ingestionendpoint",
        E = "disableExceptionTracking",
        A = "ai.device.";
      "instrumentationKey" [S = "toLowerCase"](),
      u = "crossOrigin",
      D = "POST",
      t = "appInsightsSDK",
      n = cfg.name || "appInsights",
      (cfg.name || C[t]) && (C[t] = n),
      i = C[n] || function (l) {
        var d = !1,
          g = !1,
          f = {
            initialize: !0,
            queue: [],
            sv: "7",
            version: 2,
            config: l
          };
        function m(e, t) {
          var n = {},
            i = "Browser";
          function a(e) {
            e = "" + e;
            return 1 === e.length
              ? "0" + e
              : e
          }
          return n[A + "id"] = i[S](),
          n[A + "type"] = i,
          n["ai.operation.name"] = w && w.pathname || "_unknown_",
          n["ai.internal.sdkVersion"] = "javascript:snippet_" + (
          f.sv || f.version), {
            time: (i = new Date).getUTCFullYear() + "-" + a(1 + i.getUTCMonth()) + "-" + a(i.getUTCDate()) + "T" + a(i.getUTCHours()) + ":" + a(i.getUTCMinutes()) + ":" + a(i.getUTCSeconds()) + "." + (
            i.getUTCMilliseconds() / 1e3)
              .toFixed(3)
              .slice(2, 5) + "Z",
            iKey: e,
            name: "Microsoft.ApplicationInsights." + e.replace(/-/g, "") + "." + t,
            sampleRate: 100,
            tags: n,
            data: {
              baseData: {
                ver: 2
              }
            },
            ver: 4,
            seq: "1",
            aiDataContract: undefined
          }
        }
        var h = -1,
          v = 0,
          y = [
            "js.monitor.azure.com",
            "js.cdn.applicationinsights.io",
            "js.cdn.monitor.azure.com",
            "js0.cdn.applicationinsights.io",
            "js0.cdn.monitor.azure.com",
            "js2.cdn.applicationinsights.io",
            "js2.cdn.monitor.azure.com",
            "az416426.vo.msecnd.net"
          ],
          k = l.url || cfg.src;
        if (k) {
          if ((n = navigator) && (~ (n = (n.userAgent || "").toLowerCase()).indexOf("msie") ||~ n.indexOf("trident/")) &&~ k.indexOf("ai.3") && (k = k.replace(/(\\/)(ai\\.3\\.)([^\\d]*)$/, function (e, t, n) {
            return t + "ai.2" + n
          })), !1 !== cfg.cr) 
            for (var e = 0; e < y.length; e++) 
              if (0 < k.indexOf(y[e])) {
                h = e;
                break
              }
            var i = function (e) {
              var a,
                t,
                n,
                i,
                o,
                r,
                s,
                c,
                p,
                u;
              f.queue = [],
              g || (
                0 <= h && v + 1 < y.length
                ? (a = (h + v + 1) % y.length, T(k.replace(/^(.*\\/\\/)([\\w\\.]*)(\\/.*)$/, function (e, t, n, i) {
                  return t + y[a] + i
                })), v += 1)
                : (
                  d = g = !0, o = k, c = (p = function () {
                  var e,
                    t = {},
                    n = l.connectionString;
                  if (n) 
                    for (var i = n.split(";"), a = 0; a < i.length; a++) {
                      var o = i[a].split("=");
                      2 === o.length && (t[o[0][S]()] = o[1])
                    }
                  return t[b] || (
                    e = (n = t.endpointsuffix)
                    ? t.location
                    : null,
                  t[b] = "https://" + (
                    e
                    ? e + "."
                    : "") + "dc." + (
                  n || "services.visualstudio.com")),
                  t
                }()).instrumentationkey || l.instrumentationKey || "", p = (p = p[b])
                  ? p + "/v2/track"
                  : l.endpointUrl,
                (u = []).push((t = "SDK LOAD Failure: Failed to load Application Insights SDK script (See stack for details)", n = o, r = p, (s = (i = m(c, "Exception")).data).baseType = "ExceptionData", s.baseData.exceptions = [
                  {
                    typeName: "SDKLoadFailed",
                    message: t.replace(/\\./g, "-"),
                    hasFullStack: !1,
                    stack: t + "\\nSnippet failed to load [" + n + "] -- Telemetry is disabled\\nHelp Link: https://go.microsoft.com/fwlink/?linkid=2128109\\nHost: " + (
                    w && w.pathname || "_unknown_") + "\\nEndpoint: " + r,
                    parsedStack: []
                  }
                ], i)),
                u.push((s = o, t = p, (r = (n = m(c, "Message")).data).baseType = "MessageData", (i = r.baseData).message = 'AI (Internal): 99 message:"' + (
                "SDK LOAD Failure: Failed to load Application Insights SDK script (See stack for details) (" + s + ")").replace(/\\"/g, "") + '"', i.properties = {
                  endpoint: t
                }, n)),
                o = u,
                c = p,
                JSON && (
                  (r = C.fetch) && !cfg.useXhr
                  ? r(c, {
                    method: D,
                    body: JSON.stringify(o),
                    mode: "cors"
                  })
                  : XMLHttpRequest && ((s = new XMLHttpRequest).open(D, c), s.setRequestHeader("Content-type", "application/json"), s.send(JSON.stringify(o))))))
            },
            a = function (e, t) {
              g || setTimeout(function () {
                !t && f.core || i()
              }, 500),
              d = !1
            },
            T = function (e) {
              var n = x.createElement(I),
                e = (n.src = e, cfg[u]);
              return !e && "" !== e || "undefined" == n[u] || (n[u] = e),
              n.onload = a,
              n.onerror = i,
              n.onreadystatechange = function (e, t) {
                "loaded" !== n.readyState && "complete" !== n.readyState || a(0, t)
              },
              cfg.ld && cfg.ld < 0
                ? x
                  .getElementsByTagName("head")[0]
                  .appendChild(n)
                : setTimeout(function () {
                  x
                    .getElementsByTagName(I)[0]
                    .parentNode
                    .appendChild(n)
                }, cfg.ld || 0),
              n
            };
          T(k)
        }
        try {
          f.cookie = x.cookie
        } catch (p) {}
        function t(e) {
          for (; e.length;) 
            !function (t) {
              f[t] = function () {
                var e = arguments;
                d || f
                  .queue
                  .push(function () {
                    f[t].apply(f, e)
                  })
              }
            }
          (e.pop())
        }
        var r,
          s,
          n = "track",
          o = "TrackPage",
          c = "TrackEvent",
          n = (t([
            n + "Event",
            n + "PageView",
            n + "Exception",
            n + "Trace",
            n + "DependencyData",
            n + "Metric",
            n + "PageViewPerformance",
            "start" + o,
            "stop" + o,
            "start" + c,
            "stop" + c,
            "addTelemetryInitializer",
            "setAuthenticatedUserContext",
            "clearAuthenticatedUserContext",
            "flush"
          ]), f.SeverityLevel = {
            Verbose: 0,
            Information: 1,
            Warning: 2,
            Error: 3,
            Critical: 4
          }, (l.extensionConfig || {}).ApplicationInsightsAnalytics || {});
        return !0 !== l[E] && !0 !== n[E] && (t(["_" + (
          r = "onerror")]), s = C[r], C[r] = function (e, t, n, i, a) {
          var o = s && s(e, t, n, i, a);
          return !0 !== o && f["_" + r]({
            message: e,
            url: t,
            lineNumber: n,
            columnNumber: i,
            error: a,
            evt: C.event
          }),
          o
        }, l.autoExceptionInstrumented = !0),
        f
      }(cfg.cfg),
      (C[n] = i).queue && 0 === i.queue.length
        ? (i.queue.push(e), i.trackPageView({}))
        : e();
    })({
      src: "https://js.monitor.azure.com/scripts/b/ai.3.gbl.min.js",
      // name: "appInsights",
      // ld: 0,
      // useXhr: 1,
      crossOrigin: "anonymous",
      // onInit: null,
      // cr: 0,
      cfg: { // Application Insights Configuration
        connectionString: ""
      }
    });
  </script>
  <script type="module" src="/assets/scripts/customScripts.js"></script>
  <script type="module" src="/assets/scripts/fetchResults.js"></script>

  </body>
</html>
"
`;
